plugin.js

```diff
@@ -86,7 +86,7 @@ function checkVersion (version, pluginName) {
   var fastifyVersion
   try {
     var pkgPath
-    if (require.main.filename) {
+    if (require.main && require.main.filename) {
```

Comment:
How can we get the main file using esm? I do not think the other branch of the if would work for esm, or does it?

Reply:
Actually, it does work. Before committing this I checked in a real-world project. 

Speaking of finding `the main.filename` in ESM context the only thing I found is this line `https://github.com/yargs/require-main-filename/blob/master/index.js#L5`. I don't know any other ways to get it.

Reply:
I am going to double-check this and try to add proper test too.

Reply:
OK, it works becouse we are in a `.js` file and therefore in the CJS context.
<img width="724" alt="image" src="https://user-images.githubusercontent.com/1620916/90618540-721bcc80-e210-11ea-8148-2e4076921822.png">

Reply:
And (of course) `require.main` is undefined because it was an ESM file??? :/

Reply:
>And (of course) require.main is undefined because the entrypoint was an ESM file??? :/

Yes. So we need to get the equivalent for esm

Reply:
The else branch is working in ESM. Isn't it enough?

Reply:
https://github.com/nodejs/node/pull/31875 < see this.

Reply:
Also https://github.com/nodejs/node/pull/32223

Reply:
Nice findings! I see that there are both still open and that's why I didn't found anything in the docs. :(

Are we going for `process.argv[1]` here? If so I accept suggestions on how to implement this. Maybe we could replace `require.main` entirely?

Reply:
OK I just checked all of the related issues and PR that you shared and it seems that `process.argv[1]` might be missing in the case of REPL. How does this affect our check?
Another question: is it safe for us, for the moment, to rely upon https://github.com/fastify/fastify-plugin/blob/20ebac145f395655e4fe5fefa719e80264ff6f17/plugin.js#L94 since it works in ESM?

Reply:
> since it works in ESM?

That will not work if you are using yarn2/pnp. I'm ok on unblocking this by adjusting the comment thereafter adin "as well as native ESM".

Reply:
From libtap: https://github.com/tapjs/libtap/blob/master/lib/find-main-script.js

To get the main filename under ESM context you need `process.argv[1]`, but this does not work in REPL or under `node -e` / `node -p`.  I think in the case of REPL / EVAL fastify-plugin would probably want to require.resolve relative to `path.resolve('./file-in-cwd.js')`.

In many cases just using `pkgPath = 'fastify/package.json'` will work, but if multiple copies of fastify are installed it might do the wrong thing.  Take the following structure:
```
script.mjs
node_modules/fastify - 1.x
node_modules/fastify-plugin - 1.x
node_modules/some-fastify-plugin
node_modules/some-fastify-plugin/node_modules/fastify - 2.x
node_modules/some-fastify-plugin/node_modules/fastify-plugin - 2.x
```

In this case if `some-fastify-plugin` were to have fastify-plugin check for 2.0.0 it would pass, but if we used `script.cjs` then fastify-plugin inside some-fastify-module would find the root installed fastify module and throw a version mismatch error.

This is very unlikely with npm v6 but npm v7 will install peer dependencies, so if `some-fastify-plugin` had a peer dependency on fastify ^2.0.0 this would potentially cause the above install structure.  See https://blog.npmjs.org/post/626173315965468672/npm-v7-series-beta-release-and-semver-major for details about peer dependencies.

@fox1t thank you for working on this, in the short term I had to just pin fastify-plugin to 2.0.2 in one of my projects as I didn't have a chance to work on a fix yet.

Reply:
Yea we are aware that its not going to work in the every scenario. What about something like this (I am writing this roughly just to make the point):
```js
    var pkgPath
    if (require.main && require.main.filename) {
      // We need to dynamically compute this to support yarn pnp
      pkgPath = join(dirname(require.resolve('fastify', { paths: [require.main.filename] })), 'package.json')
    } else if(process.argv[1]){ 
     pkgPath = join(dirname(require.resolve('fastify', { paths: [process.argv[1]] })), 'package.json')
    } else {
      // In bundlers, there is no require.main.filename so we go ahead and require directly
      pkgPath = 'fastify/package.json'
    }
```

Reply:
+1, add a comment that the second is for native esm.

Reply:
done